Add overview section to copilot-instructions.md (#55055)
dotnet/sdk: .github/copilot-instructions.md

Reviewer: This is now completely redundant with the other dogfood instructions above.

Author: Fixed in 6a8a0205e97

Instructions for GitHub Copilot and other AI coding agents working with the .NET SDK repository.

## Overview

This is `dotnet/sdk`, the repository for the core of the .NET SDK. It produces the
`dotnet` CLI driver and the MSBuild tasks, targets, templates, and resolvers that are
shared between the .NET CLI and Visual Studio. The build output is a complete, runnable

> `dotnet` installation laid down at `artifacts/bin/redist/Debug/dotnet`.

Author: Fixed in 55da4f09715

[README](../README.md). For the canonical build/test/debug walkthrough, see the
[Developer Guide](../documentation/project-docs/developer-guide.md).

### What the SDK does

- Provides the `dotnet` command-line driver (`dotnet build`, `restore`, `publish`, `test`,
  `run`, `watch`, etc.).
- Ships the MSBuild logic that turns a `.csproj`/`.fsproj`/`.vbproj` into a build.
- Bundles related toolsets: project/item templates, Razor/Blazor/Web/Static Web Assets
  SDKs, container publishing, file/format/watch tools, API compatibility tooling, and
  workload management.

This repo does **not** own the .NET runtime, the C#/F#/VB compilers, MSBuild itself, NuGet,
or the Visual Studio project system — those are separate repositories that flow in as
dependencies. The full product (`dotnet/dotnet` VMR) composes this repo with those.

### Architecture and major components

The CLI is a generic driver. A `dotnet <command>` is resolved one of two ways: in-box
commands implemented in `dotnet.dll` (e.g. `build`, `restore`, `test`), or `dotnet-`prefixed
executables found on the `PATH`, including installed global tools (`dotnet foo` runs

> `dotnet-foo`). The "build" commands (`build`/`restore`/`publish`/`pack`) are thin CLI

Author: Fixed in 3dae93b53a5

| `Containers/` | `dotnet publish` container image support. |
| `Dotnet.Watch/`, `Dotnet.Format/` | `dotnet watch` and `dotnet format` tools. |
| `Compatibility/` | ApiCompat / GenAPI / package validation tooling. |
| `TemplateEngine/` | CLI glue for the templating engine (engine itself is in `dotnet/templating`). |
| `Workloads/`, `Microsoft.DotNet.TemplateLocator/` | Workload manifests/installation, and locating workload-provided template packs. |
| `Layout/` | Composes the final redist `dotnet` layout. |

### Key files and directories

- `build.cmd` / `build.sh`, `test.cmd` / `test.sh`, `restore.cmd` / `restore.sh` — top-level
  entry points (Arcade-based).
- `sdk.slnx` — the full solution. Filtered solutions exist for focused work:
  `cli.slnf`, `tasks.slnf`, `containers.slnf`, `TemplateEngine.slnf`, `source-build.slnf`.
- `global.json` — pins the bootstrap SDK and Arcade versions used to build the repo.
- `Directory.Build.props` / `Directory.Build.targets` / `Directory.Packages.props` — repo-wide
  MSBuild settings and central package version management.
- `eng/` — Arcade build infrastructure, versioning (`eng/Versions.props`), and the
  `dogfood` scripts.

> - `artifacts/bin/redist/Debug/dotnet` — the built SDK; `.dotnet/dotnet` is the bootstrap SDK.

Author: Fixed in 55da4f09715

### Build, test, and dogfood

The top-level `build.cmd`/`build.sh` (and `test.cmd`/`test.sh`) wrap Arcade. The build
script restores and builds the full redist SDK; pass `-test` to also run tests. Common
switches (run `build.cmd -help` for the full list):

| Switch | Effect |
| --- | --- |
| `-c` / `-configuration <Debug\|Release>` | Build configuration (default `Debug`). |
| `-test` (`-t`) | Run tests after building. |
| `-pack` | Build installers/packages (otherwise skipped for speed). |

Arguments not directly supported by the script are passed through to MSBuild (e.g.
`/t:UpdateXlf`, `/bl` for a binlog, `/p:Property=Value`).

Canonical scenarios:

- Build the full redist SDK: `build.cmd` (Windows) or `./build.sh` (Linux/macOS). Output is
  `artifacts/bin/redist/Debug/dotnet`. First build is slow; subsequent builds are incremental.
- Build + run all tests: `build.cmd -test` / `./build.sh --test`.
- Release build: `build.cmd -c Release`.
- Run a single test project after a full build (from the build env / dogfood shell, using the
  repo-local `.dotnet`): `dotnet test test/dotnet.Tests/dotnet.Tests.csproj --filter "FullyQualifiedName~TestName"`.

>   See the [Testing](#testing) section for assembly filtering and more examples.

Author: Fixed in b62f6703a21

- `documentation/` — project docs, including the developer guide and area-specific guides.
- `template_feed/` — the in-box project/item templates.
- `test/` — test projects and `test/TestAssets/TestProjects` test inputs.

### Build, test, and dogfood

The top-level `build.cmd`/`build.sh` (and `test.cmd`/`test.sh`) wrap Arcade. The build
script restores and builds the full redist SDK; pass `-test` to also run tests. Common
switches (run `build.cmd -help` for the full list):

| Switch | Effect |
| --- | --- |
| `-c` / `-configuration <Debug\|Release>` | Build configuration (default `Debug`). |
| `-test` (`-t`) | Run tests after building. |
| `-pack` | Build installers/packages (otherwise skipped for speed). |

Arguments not directly supported by the script are passed through to MSBuild (e.g.
`/t:UpdateXlf`, `/bl` for a binlog, `/p:Property=Value`).

Canonical scenarios:

- Build the full redist SDK: `build.cmd` (Windows) or `./build.sh` (Linux/macOS). Output is
  `artifacts/bin/redist/<configuration>/dotnet` (`Debug` by default). First build is slow; subsequent builds are incremental.

> - Build + run all tests: `build.cmd -test` / `./build.sh --test`.

Reviewer: I suspect that this line will cause agents to try and run all tests. How long does it take to run all tests? Multiple hours, right? If so, we should discourage that.

Author: Fixed in 95e7f89f734

- Release build: `build.cmd -c Release`.
- Run a single test project after a full build:

>   `dotnet test test/dotnet.Tests/dotnet.Tests.csproj --filter "FullyQualifiedName~TestName"`.

Reviewer: @mthalman I know that you resolved this, but I think that this is a completely valid point. If `build.cmd`/`build.sh` behavior differs from `dotnet test`, that is unintuitive/not obvious.

>   `dotnet test test/dotnet.Tests/dotnet.Tests.csproj --filter "FullyQualifiedName~TestName"`.

Author: Fixed in 6df5fac358e

wrappers that invoke MSBuild against the SDK targets.

Major source areas under `src/`:

| Area | Purpose |
| --- | --- |
| `Cli/` | The `dotnet` driver, command implementations, and CLI utilities. `Cli/dotnet` is the entry point. |
| `Tasks/` | MSBuild tasks & targets — the heart of the build logic. `Microsoft.NET.Build.Tasks` is the primary SDK tasks assembly. |
| `Resolvers/` | MSBuild SDK resolvers (how `<Project Sdk="...">` and workloads are located). |
| `RazorSdk/`, `BlazorWasmSdk/`, `WasmSdk/`, `WebSdk/`, `StaticWebAssetsSdk/` | Web/Razor/Blazor/Static Web Assets build SDKs. |
| `Containers/` | `dotnet publish` container image support. |
| `Dotnet.Watch/`, `Dotnet.Format/` | `dotnet watch` and `dotnet format` tools. |
| `Compatibility/` | ApiCompat / GenAPI / package validation tooling. |
| `TemplateEngine/` | CLI glue for the templating engine (engine itself is in `dotnet/templating`). |
| `Workloads/`, `Microsoft.DotNet.TemplateLocator/` | Workload manifests/installation, and locating workload-provided template packs. |
| `Layout/` | Composes the final redist `dotnet` layout. |

### Key files and directories

> - `build.cmd` / `build.sh`, `test.cmd` / `test.sh`, `restore.cmd` / `restore.sh` — top-level

Reviewer: `test.cmd`/`test.sh` and `restore.cmd`/`restore.sh` really don't do much besides pass arguments to `build.cmd`/`build.sh`. Should we mention them at all if we don't tell agents to use them?

Author: Kept them listed since this is the key-files/orientation section — a newcomer sees these in the repo root and needs to know what they are. Clarified in caf2690eb83 that build.cmd/build.sh is the primary entry point and test/restore are thin wrappers that forward to it.

  `dogfood` scripts.
- `artifacts/bin/redist/<configuration>/dotnet` (`Debug` by default) — the built SDK; `.dotnet/dotnet` is the bootstrap SDK.
- `documentation/` — project docs, including the developer guide and area-specific guides.
- `template_feed/` — the in-box project/item templates.
- `test/` — test projects and `test/TestAssets/TestProjects` test inputs.

### Build, test, and dogfood

The top-level `build.cmd`/`build.sh` (and `test.cmd`/`test.sh`) wrap Arcade. The build
script restores and builds the full redist SDK; pass `-test` to also run tests. Common
switches (run `build.cmd -help` for the full list):

| Switch | Effect |
| --- | --- |
| `-c` / `-configuration <Debug\|Release>` | Build configuration (default `Debug`). |
| `-test` (`-t`) | Run tests after building. |
| `-pack` | Build installers/packages (otherwise skipped for speed). |

Arguments not directly supported by the script are passed through to MSBuild (e.g.
`/t:UpdateXlf`, `/bl` for a binlog, `/p:Property=Value`).

Canonical scenarios:

- Build the full redist SDK: `build.cmd` (Windows) or `./build.sh` (Linux/macOS). Output is

>   `artifacts/bin/redist/<configuration>/dotnet` (`Debug` by default). First build is slow; subsequent builds are incremental.

Reviewer: ```suggestion
- Full build: `build.cmd` (Windows) or `./build.sh` (Linux/macOS).
  - The script first restores a local .NET SDK to `.dotnet/dotnet`, and then builds the SDK.
  - The built SDK is output to `artifacts/bin/redist/(Debug|Release)/dotnet`.
  - The first build is slow; subsequent builds will be faster.
```

We should also mention here that `build.cmd`/`build.sh` restore the local `.dotnet/dotnet` SDK. Maybe in sub-bullets? If we do that either here or somewhere else in the file, we won't need to explain it elsewhere when we give instructions to use `.dotnet/dotnet <args>`.

Author: Fixed in bed303d444a

- Run tests: prefer targeted runs — a single test project or test (see the
  [Testing](#testing) section) and the `incremental-test` skill. `build.cmd -test` /
  `./build.sh --test` runs the **entire** suite, which is very large and takes a long time;
  avoid running the full suite for routine local or agent work.
- Release build: `build.cmd -c Release`.
- Run a single test project after a full build (in a build-env shell — see the
  [Testing](#testing) section):

>   `dotnet test test/dotnet.Tests/dotnet.Tests.csproj --filter "FullyQualifiedName~TestName"`.

Reviewer: ```suggestion
  `./.dotnet/dotnet test test/dotnet.Tests/dotnet.Tests.csproj --filter "FullyQualifiedName~TestName"`.
```

Author: Fixed in 87fd5c79302

  See the [Testing](#testing) section for assembly filtering and more examples.
- Dogfood the SDK you built (run/try it end-to-end): start a dogfood shell with
  `eng\dogfood.cmd` (Windows) or `source ./eng/dogfood.sh` (Linux/macOS). It redirects SDK
  resolution to your build so `dotnet`, `msbuild`, and Visual Studio all use it. Verify with

>   `where dotnet` / `which dotnet` (should point into `artifacts/bin/redist/<configuration>/dotnet`).

Reviewer: Copilot doesn't have a persistent shell. It's probably simpler just to tell it:

```suggestion
- Validate changes locally using the SDK you built at `artifacts/bin/redist/(Debug|Release)/dotnet`.
```

Author: Fixed in 6b7fcbfd4ed

- When creating new test projects in test/TestAssets/TestProjects, always use `$(CurrentTargetFramework)` for the `<TargetFramework>` property instead of hard-coding a specific version like `net8.0`.
- The Skip parameter of the Fact attribute to point to the specific issue link.
- To run tests in this repo:
  - First build (`build.cmd` / `./build.sh`), then load the build environment so a plain `dotnet`
    resolves correctly: dot-source `artifacts\sdk-build-env.ps1` (PowerShell), run
    `artifacts\sdk-build-env.bat` (cmd), or `source artifacts/sdk-build-env.sh` (bash).
  - For MSTest-style projects: `dotnet test path/to/project.csproj --filter "FullyQualifiedName~TestName"`

>   - For XUnit test assemblies: `dotnet exec artifacts/bin/redist/Debug/TestAssembly.dll -method "*TestMethodName*"`

Reviewer: Agents don't have a persistent shell environment. Therefore they would need to source the script every time they need to run a command. I think that it is simpler do write the following:

```suggestion
- For MSTest-style projects: `./.dotnet/dotnet test path/to/project.csproj --filter "FullyQualifiedName~TestName"`
- For XUnit test assemblies: `./.dotnet/dotnet exec artifacts/bin/redist/Debug/TestAssembly.dll -method "*TestMethodName*"`
```

Same goes for the other examples below.

Reviewer: Also for prior discussion on this exact topic in regards to agent instructions, see https://github.com/dotnet/sdk/pull/53630#discussion_r3011172541

Author: Fixed in 5269cb30bf9
